### PR TITLE
ANDAUTH-62

### PR DESCRIPTION
**Pull request by jaindiv26**
Description: Added paginable loading of users in settings, optimized this app, added "Data Saving option." and fixed SSLHandshakeException crash.

For Pagination: The API already supports the pagination, what's left was to add the support in XWiki Authenticator app. 

For Optimization of App: I've run code cleanup which removes unused imports and added in gradle.

- minifyEnabled true -> Enables code shrinking, obfuscation, and optimization for only your project's release build type. Earlier it was false.

- shrinkResources true -> Enables resource shrinking, which is performed by the Android Gradle plugin. 

For Data Saving: I have placed this option in sync settings activity. Right now this option is applicable to all users. For the implementation, I've used shared pref for storing the value.

For SSLHandshakeException crash: I saw the stack trace carefully and it happened inline 116 & 113.
at org.xwiki.android.sync.rest.XWikiHttp$2.call (XWikiHttp.java:116)
at org.xwiki.android.sync.rest.XWikiHttp$2.call (XWikiHttp.java:113)
When we look in the login method on error is implemented. But it was relogin who was calling the login method to "relogin the user". As the SSLHandshake states "Indicates that the client and server could not negotiate the desired level of security. The connection is no longer usable.". So when this error occurs there was no way that exception is handled hence the crash.

Issue links - 
https://jira.xwiki.org/projects/ANDAUTH/issues/ANDAUTH-62?filter=allopenissues
https://jira.xwiki.org/projects/ANDAUTH/issues/ANDAUTH-60?filter=allopenissues
https://jira.xwiki.org/projects/ANDAUTH/issues/ANDAUTH-63?filter=allopenissues
https://jira.xwiki.org/projects/ANDAUTH/issues/ANDAUTH-54?filter=allopenissues

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -15,7 +15,8 @@ android {
     }
     buildTypes {
         release {
-            minifyEnabled false
+            minifyEnabled true
```
Comment (InsanusMokrassar) [inline app/build.gradle:18]: Why have you enabled this?

File: app/src/main/java/org/xwiki/android/sync/bean/UserPayload.kt [hunk 2]
```diff
@@ -53,7 +53,7 @@ class UserPayload : Parcelable {
         dest.writeString(this.phone)
     }
 
-    constructor() {}
+    constructor()
```
Comment (InsanusMokrassar) [inline app/src/main/java/org/xwiki/android/sync/bean/UserPayload.kt]: This constructor was created for compatibility with Java? If not - what for?

File: app/src/main/java/org/xwiki/android/sync/activities/UserListAdapter.kt [hunk 3]
```diff
@@ -44,12 +45,14 @@ import org.xwiki.android.sync.bean.ObjectSummary
...
-    : RecyclerView.Adapter<UserListAdapter.ViewHolder>() {
+class UserListAdapter(
+    private var searchResults: List<ObjectSummary>,
+    private val context: Context
+) : RecyclerView.Adapter<UserListAdapter.ViewHolder>() {
 
 
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
-        val itemView = LayoutInflater.from(parent.context).inflate(R.layout.list_item_group, parent, false)
+        val itemView = LayoutInflater.from(parent.context).inflate(org.xwiki.android.sync.R.layout.list_item_group, parent, false)
```
Comment (InsanusMokrassar) [inline app/src/main/java/org/xwiki/android/sync/activities/UserListAdapter.kt]: Is there some requirement here to use long form?

File: app/src/main/java/org/xwiki/android/sync/activities/UserListAdapter.kt [hunk 4]
```diff
@@ -87,6 +90,7 @@ class UserListAdapter(private var searchResults: List<ObjectSummary>)
         diffResult.dispatchUpdatesTo(this)
         searchResults = listOf()
         this.searchResults = results
+        (context as SyncSettingsActivity).scrollToCurrentPosition()
```
Comment (InsanusMokrassar) [inline app/src/main/java/org/xwiki/android/sync/activities/UserListAdapter.kt]: It is not the best way. Even if you want to use casts (which is not best practice) - use `as?` to be sure that here will not be thrown some exception like ClassCastException

**Review by InsanusMokrassar — CHANGES_REQUESTED** [review]

File: app/build.gradle (hunk 1 above)
Comment (jaindiv26) [reply app/build.gradle:18]: It was in the code optimization guidelines. As I stated in the description -> 
// Enables code shrinking, obfuscation, and optimization for only your project's release build type.
minifyEnabled true

File: app/src/main/java/org/xwiki/android/sync/bean/UserPayload.kt (hunk 2 above)
Comment (jaindiv26) [reply app/src/main/java/org/xwiki/android/sync/bean/UserPayload.kt]: Removed, as it was not used anywhere.

File: app/src/main/java/org/xwiki/android/sync/activities/UserListAdapter.kt (hunk 3 above)
Comment (jaindiv26) [reply app/src/main/java/org/xwiki/android/sync/activities/UserListAdapter.kt]: Removed redundant qualifier.

File: app/src/main/java/org/xwiki/android/sync/activities/UserListAdapter.kt (hunk 4 above)
Comment (jaindiv26) [reply app/src/main/java/org/xwiki/android/sync/activities/UserListAdapter.kt]: I can put "?" safe call but the context here is non-null. So is it necessary?

File: app/src/main/java/org/xwiki/android/sync/activities/SignInViewFlipper.kt [hunk 5]
```diff
@@ -316,7 +316,7 @@ class SignInViewFlipper(
             }
 
             override fun onFailure(call: Call, e: IOException) {
-                Log.e(TAG, e.message)
+                Log.e(TAG, "No response from the server.")
```
Comment (InsanusMokrassar) [inline app/src/main/java/org/xwiki/android/sync/activities/SignInViewFlipper.kt]: Why do you sure that here may happen only absence of server response?

File: app/src/main/java/org/xwiki/android/sync/contactdb/ContactManager.kt [hunk 6]
```diff
@@ -108,11 +109,14 @@ class ContactManager(
                         if (batchOperation.size() >= 100) {
                             batchOperation.execute()
                         }
-                        contactManager.updateAvatar(
-                            resolver,
-                            contactManager.lookupRawContact(resolver, xWikiUserFull.id),
-                            xWikiUserFull
-                        )
+
+                        if (!getValue(context, "data_saving", false)) {
```
Comment (InsanusMokrassar) [inline app/src/main/java/org/xwiki/android/sync/contactdb/ContactManager.kt]: Will be better to create extension value for context, like

```
var Context.dataSaverModeEnabled: Boolean
    set(value) {
        // set in shared prefs
    }
    get() {
        getValue(this, "data_save_mode", false)
    }
```

And its usage everywhere

File: app/src/main/java/org/xwiki/android/sync/ViewModel/BaseViewModelFactory.kt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package org.xwiki.android.sync.ViewModel
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+
+class BaseViewModelFactory<T>(val creator: () -> T) : ViewModelProvider.Factory {
+    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
```
Comment (InsanusMokrassar) [inline app/src/main/java/org/xwiki/android/sync/ViewModel/BaseViewModelFactory.kt]: I am not sure that it is good idea to use `modelClass` here? I think, it is better to provide special objects for any required class

**Review by InsanusMokrassar — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/xwiki/android/sync/activities/SignInViewFlipper.kt (hunk 5 above)
Comment (jaindiv26) [reply app/src/main/java/org/xwiki/android/sync/activities/SignInViewFlipper.kt]: I'll put a more relevant message.

File: app/src/main/java/org/xwiki/android/sync/contactdb/ContactManager.kt (hunk 6 above)
Comment (jaindiv26) [reply app/src/main/java/org/xwiki/android/sync/contactdb/ContactManager.kt]: Shall I put that in the appContext?

**Pull request merged**